submariner/ipsec-cert-issuer.md

```diff
@@ -0,0 +1,114 @@
+# IPSec Certificate Issuer
+
+## Summary
+
+Submariners current security arrangements, namely using an identical pre-shared key in
+all deployments will fail security assessment in most environments which have to comply
+with a pre-existing security policy and rules on using VPNs in general and IPSEC in particular.
+
+In order to comply the minimum enhancement that needs to be implemented is the use
+of X509 certificates per Gateway. This will require the introduction of a Certificate Authority
+(CA) that's capable of signing Certificate Signing Requests (CSRs) for Gateways in the
+`ClusterSet`. This CA will be serviced by a Syncer that will act as a proxy between the Certificate
+Requester and the CA itself.
+
+## Proposal
+
+This enhancement proposes to use a `ClusterSet` wide `Certificate Authority` and a `Syncer`
+that are co-located with the Broker to service Gateway. At a high level this requires:
+
+* A Certificate Authority that can be used across the ClusterSet. Ideally this authority resides
+on the Broker node.
+* A CA_CSR_Syncer on the Broker Node.
+* A new CRD to encapsulate a Submariner CSR (towards this centralized CA), and the result (towards
+the Gateway). Gateways will fill the CRD with relevant information (a base64 encoded string of a PEM
+encoded certificate request) and publish the CR to the Kube API.
+* A GW_CSR_Syncer on the gateway Node.
+
+The general idea is that, when a Broker is installed, a CA and CA_CSR_Syncer are also installed on
+the Broker node. Then, when a Submariner gateway starts up it will launch a GW_CSR_Syncer, it will
+also create a CSR, base64 encode it and send it via a Submariner CSR CRD towards the Broker node CA.
```

Comment:
Could we use general-purpose Kubernetes CSRs instead?

Reply:
Just so we are clear are you suggesting:

1. Using the Broker Node Kubernetes API/CP as the dedicate 
ClusterSet CA and getting rid of the proposed Submariner CSR CRD? so then the syncer on the local/GW node forwards/exports CSRs to the broker node (via federator) and the rest of the process from there on out is the same (sign, publish result, publish back to GW)? OR

2. Just using the Kubernetes CSRs but still using something else for the CA such as cert-manager? 

I guess for both options 1 and 2 I need to look into the following questions:

- How would we distinguish between local (Pod/svc) CSRs vs the GW CSRs to make sure that the GW CSRs are not signed locally... and
- does the K8s CSR facility do any sort of verification of svcs/pods in the cluster before signing a cert?

The reason why I suggested cert-manager was that you can really configure different CAs behind the cert-manager API I also thought it was what ACM uses and potentially for the GW verification we may want to have an independent CA to what's provided in the scope of a local cluster.

Reply:
Using the `spec.signerName` may offer a way to distinguish whom should be signing the CSR.

Reply:
Some CSR verification in https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers.

Reply:
Right, that’s pretty much what I’m wondering. We probably still want the cert-manager API, as used in ACM, but I don’t think we need to add our own syncing mechanism. I know the addons for example have CSRs, signed using cert-manager without any other syncing mechanism, but I haven’t looked into the details — in particular, whether spoke-to-hub CSRs are supported.